### PR TITLE
util: refactor pipelines

### PR DESCRIPTION
**Pull request by chavoosh**
1- make pipeline class from utility class to a normal class
2- use pipeline class in fixed and cubic pipelines as a member variable
3- move shared functions and variables between the pipelines to the pipeline class

### CODE REVIEW

**Comment by jefft0** [comment]
This is one giant commit with 805 additions and 531 deletions, with only the comment "refactor pipelines". Did you originally make smaller commits with understandable, incremental changes and useful comments?

**Comment by chavoosh** [comment]
There is not any smaller, intermediary commit. At least I did not know how this commit could be broken down into smaller ones as the changes are tightly related. 

Technically, this commit includes the following changes:
1. define a new class, called `Pipeline` as the parent class of existing pipelines.
2. move all functions and variables that were shared between `Cubic` and `Fixed` pipelines to the parent class.
3. add `opts` parameter to `SegmentFetcher` constructor to let applications easily tune the parameters of pipelines (this could be probably a different commit, though...)

**Comment by jefft0** [comment]
I understand if you need to change the internals of the pipeline fetcher. But SegmentFetcher is a pre-existing class. Change to it should be committed separately so that a developer can understand the changes.

(I actually don't see a change to opts or to SegmentFetcher in this pull request.)

**Comment by chavoosh** [comment]
Oh my bad... I was looking at some older commits as well as the current one. You're right, the current commit does not change anything about `opts`. Please ignore the 3rd change I mentioned in my previous comment.

So, this commit is only about the pipelines (without touching any other part of the source code). If still you think this commit needs go through some changes, please let me know...

**Comment by jefft0** [comment]
You say "define a new class, called Pipeline as the parent class of existing pipelines." But the current master already has the class Pipeline. Maybe you mean that you turned it into a normal class instead of a utility class with only static methods?
https://github.com/named-data/ndn-js/blob/master/js/util/pipeline.js

You say that Pipeline is the parent class of existing pipelines. But, for example, the class PipelineCubic does not use Pipeline as a parent class. Instead it has a member variable pipeline which has an instance of Pipeline. 

Can you please update the description to this pull request to clearly state what it changes?

File: js/util/pipeline.js [hunk 1]
```diff
@@ -74,7 +102,48 @@ Pipeline.ErrorCode = {
...
+
+/**
+ * Make an Interest with proper name
+ * @description If baseInterest's name contains version number it will be used, otherwise the
+ *              discovered version number will be used. Then @param segNo is appended to the name.
+ *              If no version number is available, we send back a copy of baseInterest.
+ */
+Pipeline.prototype.makeInterest = function(segNo)
+{
+  var interest = new Interest(this.baseInterest);
```
Comment (jefft0) [inline js/util/pipeline.js]: To use the Interest class in node.js, at the top of the file please add 

    var Interest = require('../interest.js').Interest;

**Comment by chavoosh** [comment]
> You say "define a new class, called Pipeline as the parent class of existing pipelines." But the current master already has the class Pipeline. Maybe you mean that you turned it into a normal class instead of a utility class with only static methods?

You're right. I should have used a better description. As you said, here this commit turns the `Pipeline` class from a utility class to a normal class.

> You say that Pipeline is the parent class of existing pipelines. But, for example, the class PipelineCubic does not use Pipeline as a parent class. Instead it has a member variable pipeline which has an instance of Pipeline.

Agina, you're right. Wrong explanation... `Fixed` and `Cubic` pipelines use `Pipeline` class as a member variable. 
Quick question: "does using `Pipeline` as parent class change the performance? If it does, I will update the code."


Jeff, thanks for reviewing the code.

**Comment by jefft0** [comment]
No, I don't think it will change the performance.

**Comment by chavoosh** [comment]
> No, I don't think it will change the performance.

So, if you do not see any reason to refactor the code to extend pipelines based on `Pipeline` class, let's keep the code as it is.

File: js/util/pipeline.js (hunk 1 above)
Comment (jefft0) [reply js/util/pipeline.js]: If you make this change, I think we're ready to merge.

**Comment by chavoosh** [comment]
Can this be merged?

**Comment by jefft0** [comment]
Yes, except for the comment I made about needing a require statement in pipeline.js. GitHub should send you a notification when someone makes a comment on a line in a file. Do you get these notifications?

**Comment by chavoosh** [comment]
> Yes, except for the comment I made about needing a require statement in pipeline.js. GitHub should send you a notification when someone makes a comment on a line in a file. Do you get these notifications?

Yes, I do. I forgot to apply it. I will push a new commit in a few minutes. Also, I want to run the commit message by you to make sure it is fine...

```
util: refactor pipelines 
1- make pipeline class from utility class to a normal class
2- use pipeline class in fixed and cubic pipelines as a member variable
3- move shared functions and variables between the pipelines to the pipeline class
```

**Comment by jefft0** [comment]
Yes, the commit message looks good.

**Comment by chavoosh** [comment]
I have pushed a new commit.

**Pull request merged**